staging/src/k8s.io/apiserver/pkg/cel/ip.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"math"
+	"net/netip"
+	"reflect"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// IP provides a CEL representation of an IP address.
+type IP struct {
+	netip.Addr
+
+	// Raw is the original string representation of the IP address.
+	// This is used so that we can compare the original string representation
+	// with the parsed IP address.
+	Raw string
+}
+
+var (
+	IPObject        = decls.NewObjectType("kubernetes.IP")
+	ipAddrTypeValue = types.NewTypeValue("kubernetes.IP")
+	IPType          = cel.ObjectType("kubernetes.IP")
+)
+
+// ConvertToNative implements ref.Val.ConvertToNative.
+func (d IP) ConvertToNative(typeDesc reflect.Type) (interface{}, error) {
+	if reflect.TypeOf(d.Addr).AssignableTo(typeDesc) {
+		return d.Addr, nil
+	}
+	if reflect.TypeOf("").AssignableTo(typeDesc) {
+		return d.Addr.String(), nil
+	}
+	return nil, fmt.Errorf("type conversion error from 'IP' to '%v'", typeDesc)
+}
+
+// ConvertToType implements ref.Val.ConvertToType.
+func (d IP) ConvertToType(typeVal ref.Type) ref.Val {
+	switch typeVal {
+	case ipAddrTypeValue:
+		return d
+	case types.TypeType:
```

Comment:
Do you possibly want to consider conversion to string? I believe this will help with the `string.format` call

Reply:
Rather that produce the string type name, can you convert back to a string representation?

Reply:
Ahh right I see, I looked at the upstream code for inspiration on this and ended up at the `type` type, I should have looked at a few more types to see what they did, can fix